### DISCUSSION TITLE
Block rendering based on context (universal blocks)

### DISCUSSION THREAD

**Discussion by joshuatf** [woocommerce/woocommerce#39775] [General]
The idea of a "universal block" is one where we can add a block to a number of different contexts and it reacts appropriately to the context it's placed in.  For example, the same product title block could be added to the product editor form and render the input field, added to the single product page to render the title, and added to a product list to render the smaller version of the title, potentially linked to the product page.

## Implementation

At it's core, the idea is that a block will render differently based on it's context:

```js
function Edit() {
	const { context } = useProduct();

	if ( context === 'form' ) {
		return <Field />;
	}

	if ( context === 'list' ) {
		return <SmallTitle />;
	}

	return <MainBlock />;
}
```

This could be simplified to a wrapper for `Edit` and `Save` methods and some API for establishing when each should render, offloading responsibility of determining context and repitition between render conditions.  This is just an idea and not a fully baked one, but as an example:

```jsx
// Woo block registration utils
function getEditRender( edit ) {
	if ( typeof edit === 'function' ) {
		return edit;
	}

	return () => {
		const { context } = useProduct();
	
		if ( editBlocks.hasOwnProperty( context ) ) {
			const Component = edit.hasOwnProperty( context );
			return <Component />;
		}

		const DefaultComponent = edit.default;
		return <DefaultComponent />;
	}
}

function registerWooBlock( { name, edit } ) {
	const EditComponent = getEditRender( edit );
	const metadata = { name, edit: EditComponent };
}

// Block registration
registerWooBlock( {
	name: 'blockName',
	edit: {
		form: Field,
		list: SmallTitle,
		default: MainBlock
	}
} );
```

## Potential issues

The main problem I see in doing this is this sets us up to use a single `block.json` file for a universal block.  And at first glance this is a nice reduction in code, but the block metadata is likely to vary based on where a block is used.  This is especially true when comparing a "field" block (used in the product editor) to a block that is used to visually edit and render respective content on the frontend.

The following block metadata properties are common culprits where I think we'd see differences depending on the block context:

* `attributes` - The attributes used for the same block vary depending on context.  I think this can be reconciled by trying to conform the properties and adding all the properties that need to be used in each context, but this could get messy.
* `supports` - The [supports property](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/) determines what tools a block supports.  Because many of the field-based blocks don't support toolbars, alignment, or other properties that would be needed in a visual editing context, this seems impossible to resolve in a single block file.
* `parent|ancestor` - These properties require that a block is a descendant of another block.  This will vary between contexts and does not appear to be something we can resolve in a single block file.

I believe what this ultimately means is that we need multiple `block.json` files still for a single "universal block."  This obviously reduces some of the perceived benefits of this approach.

## Filtering block settings

One potential workaround to needing duplicate block metadata for the same block would be to use the filter in blocks:

https://github.com/WordPress/gutenberg/blob/14600f31582ba0ebe7343130d864ac6dba1d73db/packages/blocks/src/store/actions.js#L59-L64

While client-side filters do have issues with timing, Gutenberg has a `__experimentalReapplyBlockTypeFilters` action in the block data store that is used to reapply settings once an editor is ready.  This could be run after the block editor and context are ready.

```jsx
// Woo block registration utils
function getEditRender( edit, metadata, contextMetadata ) {
	return () => {
		const { context } = useProduct();

		useEffect( () => {
			const contextMetadata = contextMetadata[ context ] || {};
			addFilter(
				'blocks.registerBlockType',
				'woocommerce/set-block-contextual-settings',
				() => {
					return {
						...metadata, // Default metadata
						...contextMetadata, // Context-specific metadata
					}
				}
			);

			// This may need to be moved elsewhere to avoid an On^2 scenario when reapplying settings to blocks.
			dispatch( blocksStore ).__experimentalReapplyBlockTypeFilters();
	
			return () => {
				removeFilter( ... );
			}
		}, [ context ] );

		// Render the component based on context...
	}
}

function registerWooBlock( block ) {}

// Block registration
import defaultBlockMetadata from './block.json';
import formBlockMetadata from './form/block.json';
import listBlockMetadata from './list/block.json';

registerWooBlock( {
	...defaultBlockMetadata,
	contexts: {
		form: formBlockMetadata,
		list: listBlockMetadata,
	},
} );
```

In short, this would apply block settings at runtime based on the context the block is in.  There would be a primary block file and several other metadata files could be used to only override specific settings based on context:

```json
// block.json
{
	"name": "blockName",
	"version": "1.0.0",
	"title": "Some block",
	"category": "woocommerce",
	"keywords": [ "WooCommerce" ]
}

// form.json
{
	"parent": [ "woocommerce/form-section" ]
}
```

## Determining context

In terms of determining context and the product/order or other respective entity we're currently using, we have a few options at our disposal.

### Reliance on the entity data store

The entity data store already provides a great way to see the current entity, both in its persisted and edited state.

Because of the built-in functionality around entity data stores, I'm a fan of using this data store wherever possible.

This [entity config](https://github.com/WordPress/gutenberg/blob/14600f31582ba0ebe7343130d864ac6dba1d73db/packages/core-data/src/selectors.ts#L243-L251) might be a potential candidate for storing context as well, but I'll have to dig into this to see if that's a viable solution and if determining context makes sense based on the entity.

### Pass context via block context

The `<BlockContextProvider>` provides a [built-in way to pass context to blocks](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#context).

This may provide a bit more flexibility than trying to use the entity data store and has the advantage of not needing to export/import hooks or custom context to determine context in third party blocks.

### Custom context providers

And lastly, we can always create our own context providers.  This has been done in WooCommerce Blocks to allow grabbing product data in blocks using the [ProductDataContextProvider](https://github.com/woocommerce/woocommerce-blocks/blob/f8ce88888bd7bc80629c0849cdd5b22b73d843b2/assets/js/shared/context/product-data-context.tsx#L85-L105).

This option has the most flexibility in allowing us to use context outside of the scope of blocks if needed, but also comes with a few additional maintenance costs to us.

It would be great if @woocommerce/rubik could provide some insight to the reasons behind using this over the entity data store and block context as this may help guide our decisions in the best approach.

## Benefits of contextual blocks

What are the benefits to using Universal blocks?  As I see it, the main advantages are:

* Consolidation of related block code in a single place
* Reducing some level of duplication around attributes and other block metadata used in each blocks
* Reuse of utils that reside in a specific block

I think @nerrad may be able to speak to other benefits not mentioned above so that we can determine whether or not this an approach worth pursuing.  There are some obvious technical challenges and complexities added by going this route, so we need to make sure the positives outweigh those challenges.

**Comment by dmsnell** [comment]
This reminds me somewhat of what Tumblr does with their media placeholder. When inserting media there's a generic block. When uploading a file it then calls `replaceBlock()` to insert the specific kind.

I'm curious what you've already tried as a means for reuse of logic, which you list as the benefits of this approach.

> The main problem I see in doing this is this sets us up to use a single block.json file for a universal block. And at first glance this is a nice reduction in code, but the block metadata is likely to vary based on where a block is used. This is especially true when comparing a "field" block (used in the product editor) to a block that is used to visually edit and render respective content on the frontend.

This does seem like an important aspect to consider, and one of the reasons that pushing to de-duplicate code often backfires on us by deeply-coupling together things that behave differently.

I wonder if there are opportunities to break apart behaviors horizontally into functions that can be used and composed together in these similar block types without mandating that they all fit into the same footprint.

How many specific variations of a universal block are we discussing? I would imagine that the story is different if we are talking about four or five variations vs. potentially hundreds of variations coming from user-generated content.

**Reply by joshuatf** [reply]
Thanks for the feedback, @dmsnell!

>I'm curious what you've already tried as a means for reuse of logic, which you list as the benefits of this approach.

This is more of an indirect benefit that comes with consolidation.  Part of the issue we face right now is that "pieces" of the universal block live in separate repos.  For example, the product name/title exists in WooCommerce under `@woocommerce/product-editor` and the more visual forms of this block exist in WooCommerce Blocks.

Sharing logic between the two is possible currently but relies on exports and opens up potentials for unexpected behavior if we rely on external dependency extraction.

The argument could be made that we could house these in one place without needing to make them a universal block or that we could move logic into a dedicated package (e.g., `formatCurrency` exported from `@woocommerce/currency`) which is probably the most ideal long-term solution.  Universal blocks just provide incentive to group similar items together, reducing the short-term need for exports or additional packages.

>I wonder if there are opportunities to break apart behaviors horizontally into functions that can be used and composed together in these similar block types without mandating that they all fit into the same footprint.

I'm leaning towards this as well.  This is certainly possible and done to some degree already in WooCommerce Blocks.  It may mean that admin related blocks and frontend blocks continue to stay separate as there's such a large divergence in both render and metadata.

@nerrad is back next week and may be able to shed some light on any additional benefits I may have missed as I know he was a proponent of this idea.

>How many specific variations of a universal block are we discussing? I would imagine that the story is different if we are talking about four or five variations vs. potentially hundreds of variations coming from user-generated content.

That's a good question.  I think the same block may exist in the realm of 5-10 different places on average, but this is really going to vary from block to block.  In the case of the product title:

* Product edit page
* Admin product list page
* Frontend product list page
* Single product
* Order item
* Cart item

That's not an exhaustive list, but also worth noting that I think many of these could be merged into categorical types.  Cart items, order items, frontend product lists, and admin product lists may all belong to a "list" context and share similar functionality.  I think this is superior to having context for each individual screen, especially once we consider third party developer context as well.

**Reply by dmsnell** [reply]
> The argument could be made that we could house these in one place without needing to make them a universal block…which is probably the most ideal long-term solution.

sounds like a good start in any case. it can be so hard to build valuable abstractions when trying to consolidate code due to the way we focus on removing duplication. if you go the universal block route, though, it would be good to report on how it goes and ask others how they like it.

> there's such a large divergence in both render and metadata.

this is what I would anticipate is one of the biggest risks for such a consolidation. we could start out optimistic that things that look alike are truly alike, and then realize that the details which differentiate them make the universal wrapper complicated enough that's mostly a now-confusing middleman.

hope it turns out well though. thankfully we can go pretty far along the way by building reusable bits and sharing those. maybe those can help us understand what needs a universal block would have.

**Reply by nerrad** [reply]
Some intent behind the "universal block" idea (and a bit of a ramble):

- limit duplication of code.
- a single block to represent the content across any context. 
- simplifying the registration surface area for a block so that the block author doesn't need to be aware of when and where to register it. To a degree, the block could be agnostic about its editor context - instead exposing what to render in the relevant _*user* context_. 

The challenge that seems to be surfacing here is that the product editor is rendering a user context that is explicitly ignoring any frontend visual feedback to favour primarily data entry which is a tangent from how blocks have been usually used across editors. In some ways, the render model for product editor blocks is more akin to what might be implemented within `<InspectorControls/>`.

Within Woo owned/published blocks, it definitely seems more prudent to follow the composition approach however I'm also interested in the devex for 3PDs developing Woo Blocks and the registration process around those blocks. Aside from the choices of what is rendered for the user interactions, should developers have to care about what editor a block is registered for? Currently, they don't (for site or post editors).

**Reply by joshuatf** [reply]
> In some ways, the render model for product editor blocks is more akin to what might be implemented within <InspectorControls/>.

💯  I think this is a really nice way of summing up the current differences in editing experience between these two types of editors.

>Aside from the choices of what is rendered for the user interactions, should developers have to care about what editor a block is registered for?

I think it's not overly likely they would need to be concerned with registration within something like the product area which has a focus on data entry and has a much more locked down template where merchants cannot directly add blocks.  That said, devs who do want to add blocks to that editor would have greatly simplified registration as you mentioned in your bullet points.

As far as other areas go, I can definitely see benefit from blocks being context aware.  A third party block that shows a price in different currencies may want to render differently depending on the context in which they're added.

I think if that's something worthwhile, the next steps would be to:

* See if we can agree on predefined contexts and naming conventions.  E.g., `product/list`, `order/table`, etc.
* Determine if different block configs are needed for each context.  Even excluding the more data-focused field blocks, I think the answer to this is "yes" if we want to continue supporting `parent` and `ancestor` properties for each of those contexts.

I'd love some more buy-in from @woocommerce/mothra and @woocommerce/rubik before spend too much time on a POC to see if there are any concerns or if this is something you'd like to see supported.

**Reply by mattsherman** [reply]
I like the high-level concept of a universal block, where I, as a dev can implement thing in a single place and have it automatically use the "correct" implementation based on context, if the number of different contexts to be supported is small. I'd be curious to see what a POC would look like for something currently used in all the different contexts we envision.

**Reply by louwie17** [reply]
> This does seem like an important aspect to consider, and one of the reasons that pushing to de-duplicate code often backfires on us by deeply-coupling together things that behave differently.

It might be worth having a use case of one block in all the different contexts that @joshuatf briefly mentioned in the comment above and see what exactly is different in the attributes, support settings, Edit component and Save component.
I feel like the product editor block would be the outlier having the biggest differences in attributes and visual representation. I would be more apt to keep them different if that is truly the case, and just re-use any of the same components/hooks through a shared library.

If we still want the exact same registration process, we can make use of `supports` and add an additional editorContext. So for potentially the few cases where attributes stay the same across all editor contexts users have the option to just register the single block.

**Reply by joshuatf** [reply]
>If we still want the exact same registration process, we can make use of supports and add an additional editorContext.

Love this idea @louwie17!  This could be useful for making sure these blocks don't show up in certain contexts too and add more ability for us to filter these blocks during registration.

**Reply by joshuatf** [reply]
> if the number of different contexts to be supported is small

It will definitely be a challenge to nail down a precise set of predefined contexts.  Do you have a number or range in mind that is "small" @mattsherman?

**Comment by mattsherman** [comment]
> > if the number of different contexts to be supported is small
> 
> It will definitely be a challenge to nail down a precise set of predefined contexts. Do you have a number or range in mind that is "small" @mattsherman?

"small" to me feels like 3-4 contexts.

**Comment by mattsherman** [comment]
> I feel like the product editor block would be the outlier having the biggest differences in attributes and visual representation

I agree.

Conceptually, I wonder if we should have had the product editor implemented such that the product editor that was rendered was actually the "view" mode and the "edit" mode was some (unimplemented) future mode that would allow the form/template to be customized via UI (a "form editor"). I don't know if that would help the implementations converge any in a universal block concept, but it does feel like it would conceptually help a little. Maybe not though... this idea might just muddy the waters. If so, let's pretend I never brought it up! 😄

**Reply by joshuatf** [reply]
I like the creative thinking!  Kind of in the same vein, I've been wondering if we should have made the product editor entirely using the sidebar and left of the edit/save views for future-proofing any visual editor we may want to implement.

Essentially we would just expand the sidebar portion into the entire width of the page for now and use those form elements to build the product:
<img width="353" alt="Screen Shot 2023-09-07 at 4 36 46 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/a0ab930c-cecf-41e9-a4b9-a4686aa0fca5">

This solves a lot of the problems we have around trying to turn off editor features (insertion, toolbars, locking, etc) while also leaving the editor canvas itself free to customize a product if we decide to add that feature in the future.  But obviously that means backtracking on some of the work we've done and rethinking form extensibility a bit.

**Reply by nerrad** [reply]
> I've been wondering if we should have made the product editor entirely using the sidebar and left of the edit/save views for future-proofing any visual editor we may want to implement.

That's been on my mind too.

**Reply by louwie17** [reply]
> I've been wondering if we should have made the product editor entirely using the sidebar and left of the edit/save views for future-proofing any visual editor we may want to implement.

Interesting thought, that would be interesting to explore.
Also out of curiosity ( in the case we do explore this ), it looks like the sidebar implements the fields [more manually](https://github.com/WordPress/gutenberg/blob/60a6c2df2794cc90f950e8a05cd266fca33fc56f/packages/edit-post/src/components/sidebar/post-status/index.js#L48) with the use of SlotFills, exactly like our old implementation. In this case if we wanted to continue using templating we did have to make that work within the sidebar, although this is likely not a huge amount of work given we have done a lot of the heavy lifting already. 

I do think it will be worth building a light weight version of the template block renderer anyway, that doesn't include the insertion, toolbars, and locking that we had to circumvent.
Going a long with that, we did almost have to create a base block renderer, that includes a base config.
We then have Gutenberg extend from that (adding the required `edit` & `save` configs for example ), the product editor would also extend from this, depending on how many modifications we had to make.

**Comment by joshuatf** [comment]
# Proof of concept

I've created a POC that shows contextual blocks in action in https://github.com/woocommerce/woocommerce/pull/40077.  This example shows the same block being rendered in different ways depending on where it's placed.

Overall, I found this to be easier than I expected to implement and the `registerWooBlock` providers wrappers and context that are useful beyond just contextual blocks.  However, there are a few caveats to this solution working.


<img width="725" alt="Screen Shot 2023-09-07 at 2 12 15 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/f3b2539f-db2a-46ee-85cd-85865e2d3d29">

## Block settings consistency

I mentioned above that we may be able to tap into the `blocks.registerBlockType` hook to filter settings.  While that's true, this would not work well if a block could exist in multiple contexts in a single editor since settings are established when the block is first registered.  We also don't have access to the block context at the point of registry, making this impossible to accomplish based on context alone.

The good news is that I think for the most part, the settings for a block can remain pretty consistent and settings which would vary from context to context, such as `parent` or `ancestor` accept an array of block types so that we can cover all the use cases.

The `supports` property has a little more potential for inconsistency, but the primary known culprit as of writing is with the support properties around disabling the toolbar and locking content.  Luckily this can also be locked via parent blocks in that editor without the use of `supports`, making it feasible to use the exact same `block.json` metadata for each of the contexts.

### Metadata options

#### Singular block.json

The simplest solution is to accept that contextual blocks need to maintain a single metadata file.  If this works for all known use cases, this may be a good path forward.

#### Added settings filter

If this is a dealbreaker, we could push for a filter upstream in Gutenberg to allow modifications of block settings at runtime.

However, while this might be a useful feature for the community, it would be a significant change to the existing Gutenberg codebase that uses `getBlockType` to retrieve block settings throughout the app (with 358 instances of its use that would need to be audited).  We would want to replace this with some type of `useBlockSettings` hook that allowed filtering and provided not only the block, but current editor context.

## Narrowing context options

I'm not sure that we can create a finite list of all possible contexts for all possible scenarios.  At best, we can come up with very vague categories for contexts.  Here are some possible options for categorizing context options:

### Generic categories

I think this is the best chance we have at making categories that cover most blocks.  These semantics are a WIP, but here are pretty common places the same block might be used:

* **Page** - The entity's specific page. E.g. the product name used in the header on a single product page.
* **Item** - A singular item, often used in a loop.  E.g., The product name used in the grid for the product shop page
* **List** - The item when displayed in list format.  E.g, a product name used in the cart items list
* **AdminList** - Similar to above, but specific to admin lists.  The difference being the href location.
* **Form** - The form component used in the product editor.

### Layout-based categories

Another option is to create categories around the layout of the site.

* **Header** - Used in the header area.
* **Body** - Used in the body of the page.
* **Element** - This might be a card or some type of standalone element used within a loop.  This could probably be improved semantically.
* **Table/List** - Items used within any table or list.
* **Admin** - Used anywhere in the admin context
* **Form** - Used anywhere in a form context.

### Specific categories

One last and more open-ended approach would be to not try and pin down categories that work across the board and accept that each block can have its own context needs.  A product name might be used in the following scenarios:

* **Shop** - The shop page/lop of products.
* **Cart** - Anything that resides in the cart container.
* **Admin products** - The list of products in the admin area
* **Order details** - The order received page with order information.
* **Admin Form** - The product editor.
* Etc..

The disadvantage to this is discoverability of contexts.  We'd want a way to document and track contexts used around the site and make sure that blocks had defaults for most high-level containers.

The biggest advantage is that way we can create very specific blocks in certain scenarios that are extremely context aware.  Also worth noting that this currently works out-of-the-box with the current POC.

## Next steps

In order to move forward with this, we have a couple of decisions to make.

1. Does this concept add enough value?  The wrapper that came out of this has many use cases, however this question is more about block context specifically.
2. Do we think we can get away with a singular block metadata file or do we need multiple?  I'm leaning towards being able to use a single file, but am curious if we have known conflicts that make this unfeasible. 
3. Which category structure works the best?  Layout-based and specific categories stand out as the best options to me, but I'm thinking if we try hard, we can come up with examples that break the layout-based option.

**Reply by dinhtungdu** [reply]
> Does this concept add enough value?

Yes, I think having multiple representations of the same data make a lot sense to me. Doing so, we can even build a visual product editor that based on product template, maybe even a frontend editor. I'm sold just by that idea 💯 .

> Do we think we can get away with a singular block metadata file

Currently, yes, I do. But we also should try hard to break that assumption.

> Which category structure works the best?

The last one sounds best to me. A block can define which context it support. It's also more future-proof as well. But yeah, it comes with the cost of managing,

**Reply by nerrad** [reply]
I think this concept has value. However, given `block.json`, has a defined, published, schema this might be one area where we could run into conflicts (especially as it is versioned and updates might break our implementations). 

To address that potential issue, should `block.json` be reserved for the native WP editors and we introduce a separate spec for any custom Woo editor instances? Feasibly `registerWooBlock` would be a wrapper that extends the `registerBlock` API and receives the additional metadata for any contexts. The wrapper handles registering for the different contexts automatically.

```js
/**
 * `wp` is the metadata value for registering blocks on the default WP editors (post and site).
 * `form` is the metadata value for registering blocks in the product editor (or expanded form context, which might also be applicable in the future for embedded edit views elsewhere.
 */
registerWooBlock( {
    name,
    wp: blockJsonMetadata,
    form: { ...blockJsonMetadata, edit: formComponent }
} );
```

That way extensions could recompose their components however they want and `registerWooBlock` simply takes care of registering the appropriate block for the relevant routes. This structure also allows for:

- keeping the schema of `block.json` as is with extensions implementing overrides on necessary metadata properties.
- optionally in the future, providing enhanced metadata schemas for these specialized editors (i.e. `form.json`) but we don't have to do that right away.

I'd also avoid getting too granular with categories for the contexts here and keep it initially really simple. Only breaking out a new context top-level property when it is clear it is needed.

**Reply by joshuatf** [reply]
Thanks for the input, @dinhtungdu!

>It's also more future-proof as well

This is the point that stands out to me also.  It's difficult to imagine covering all potential contexts with a finite set of categories.

**Reply by joshuatf** [reply]
Good thoughts, @nerrad!

>Feasibly registerWooBlock would be a wrapper that extends the registerBlock API and receives the additional metadata for any contexts. The wrapper handles registering for the different contexts automatically.

I'm not sure if you've already seen the POC, but this is similar to how it handles things, though the context renders are intermingled with the normal block registration:

```js
registerWooBlock( {
    metadata: blockMetadata,
    edit: Edit, // This can be a single component as traditionally done with `registerBlockType` or an object with key value pairs for context => component
} );
```

>form: { ...blockJsonMetadata, edit: formComponent }

I like the separation of contexts from the default block metadata, though I'm wondering how much use we can squeeze out of this.  I think the renders may be the only place we can be context aware, limiting the differences in contexts purely to render and not other contextual block metadata.

I moved away from having any other differences in block metadata for some of the reasons you mentioned, but also because we would not be able to have 2 different contexts on a single page where blocks respected their provided metadata since we can only officially register a single set of block settings.

>I'd also avoid getting too granular with categories for the contexts here and keep it initially really simple. Only breaking out a new context top-level property when it is clear it is needed.

Does this mean you're leaning towards generic context categories?  If so, I think limiting to a small set of categories is a good call 👍 .

**Reply by nerrad** [reply]
> I moved away from having any other differences in block metadata for some of the reasons you mentioned, but also because we would not be able to have 2 different contexts on a single page where blocks respected their provided metadata since we can only officially register a single set of block settings.

The different renders (i.e. `form.edit` vs `wp.edit`) in my example are intentional. The idea is, that if a specific editor might use both contexts (eg the product editor description modal using the `wp.edit` block), _both_ renders would be registered as separate blocks (sharing the base settings). Admittedly this idea is very raw and there are practical implications that might make it infeasible.

> Does this mean you're leaning towards generic context categories?

If you mean the categories you have listed there (Page, List etc), I think that's still too specific. For example, the distinction between a block rendered in the context of product collection or a single product page is something that shouldn't need explicit registration. If _needed_ the block should be able to infer its context from ancestors (even then that's not really needed in most cases).

**Reply by joshuatf** [reply]
>both renders would be registered as separate blocks (sharing the base settings)

Ah I see.  I had considered this as well, but mostly ruled it out for a couple reasons:

* I think if they are separate blocks with separate IDs, we may run into confusion with templates knowing which version of the block to use.  I think the name of the block needs to remain consistent to get most of the contextual benefits.
* If there is a partial block metadata that is meant to "inherit" pieces of the primary block metadata file, the current state of WordPress block registration is going to hit issues parsing the "sub-blocks."

I'm sure we can work around these items with some hacks, but it starts to feel like we're losing some of the benefits of the shared block data and adding a lot of complexity which could make debugging challenging.

>If you mean the categories you have listed there (Page, List etc), I think that's still too specific.

Do you have examples of what you have in mind?

**Reply by nerrad** [reply]
> Do you have examples of what you have in mind?

The only one I think is distinctive right now from a block registration/surfacing perspective is "form".

**Reply by joshuatf** [reply]
That's another interesting way of categorizing the contexts.  I don't disagree with that categorical structure at all, but until we have more categories, it does make me wonder:

* Should we really begin considering placement of form elements into the sidebar as opposed to treating them as classic blocks?
* Similar to your above statement regarding inferring render via ancestry, I think we could also do the same with form based elements.
